File: src/utils/productValidation.ts
import { ProductBase, ProductFormData, ProductValidationErrors } from '../types/product';

/**
 * Validates a product form submission
 * @param product - The product data to validate
 * @returns An object containing validation errors, if any
 */
export const validateProduct = (product: ProductFormData): ProductValidationErrors => {
  const errors: ProductValidationErrors = {};

  // Title validation
  if (!product.title.en || product.title.en.trim() === '') {
    errors.title = errors.title || {};
    errors.title.en = 'Product title in English is required';
  } else if (product.title.en.length < 3) {
    errors.title = errors.title || {};
    errors.title.en = 'Product title must be at least 3 characters';
  }

  if (!product.title.es || product.title.es.trim() === '') {
    errors.title = errors.title || {};
    errors.title.es = 'Product title in Spanish is required';
  } else if (product.title.es.length < 3) {
    errors.title = errors.title || {};
    errors.title.es = 'Product title must be at least 3 characters';
  }

  // Description validation
  if (!product.description.en || product.description.en.trim() === '') {
    errors.description = errors.description || {};
    errors.description.en = 'Product description in English is required';
  }

  if (!product.description.es || product.description.es.trim() === '') {
    errors.description = errors.description || {};
    errors.description.es = 'Product description in Spanish is required';
  }

  // Price validation
  if (product.price === undefined || product.price === null) {
    errors.price = 'Price is required';
  } else if (isNaN(product.price) || product.price < 0) {
    errors.price = 'Price must be a positive number';
  }

  // SKU validation
  if (!product.sku || product.sku.trim() === '') {
    errors.sku = 'SKU is required';
  } else if (!/^[A-Za-z0-9-_]+$/.test(product.sku)) {
    errors.sku = 'SKU can only contain letters, numbers, hyphens, and underscores';
  }

  // Category validation
  if (!product.categoryId) {
    errors.categoryId = 'Category is required';
  }

  // Inventory validation
  if (product.inventory.quantity === undefined || product.inventory.quantity === null) {
    errors.inventory = errors.inventory || {};
    errors.inventory.quantity = 'Inventory quantity is required';
  } else if (isNaN(product.inventory.quantity) || product.inventory.quantity < 0) {
    errors.inventory = errors.inventory || {};
    errors.inventory.quantity = 'Inventory quantity must be a positive number';
  }

  // Images validation
  if (!product.images || product.images.length === 0) {
    errors.images = 'At least one product image is required';
  }

  return errors;
};

/**
 * Checks if a product has validation errors
 * @param errors - The validation errors object
 * @returns True if there are validation errors
 */
export const hasValidationErrors = (errors: ProductValidationErrors): boolean => {
  return Object.keys(errors).length > 0;
};

/**
 * Generates a URL-friendly slug from a product title
 * @param title - The product title
 * @returns A URL-friendly slug
 */
export const generateProductSlug = (title: string): string => {
  return title
    .toLowerCase()
    .replace(/[^\w\s-]/g, '') // Remove special characters
    .replace(/\s+/g, '-') // Replace spaces with hyphens
    .replace(/-+/g, '-') // Replace multiple hyphens with single hyphen
    .trim(); // Trim spaces from start and end
}; 

File: backend/app/utils/product_validation.py
import re
from datetime import datetime

def generate_product_slug(title: str) -> str:
    """Generate a URL-friendly slug from a product title"""
    # Convert to lowercase and replace spaces with hyphens
    slug = title.lower().strip()
    
    # Replace spaces and special characters with hyphens
    slug = re.sub(r'[^a-z0-9\s-]', '', slug)
    slug = re.sub(r'\s+', '-', slug)
    slug = re.sub(r'-+', '-', slug)
    
    # Remove leading/trailing hyphens
    slug = slug.strip('-')
    
    # Ensure it's not empty
    if not slug:
        slug = f"product-{int(datetime.now().timestamp())}"
    
    return slug

def validate_sku(sku: str) -> bool:
    """Validate SKU format"""
    if not sku or len(sku) < 3:
        return False
    
    # Allow letters, numbers, hyphens, and underscores
    return bool(re.match(r'^[A-Za-z0-9_-]+$', sku))

def validate_price(price: float) -> bool:
    """Validate price is positive"""
    return price > 0

def validate_product_data(data: dict) -> tuple:
    """Validate product data and return (is_valid, error_message)"""
    
    # Check required fields
    required_fields = ['title', 'price', 'sku', 'category_id']
    for field in required_fields:
        if field not in data or not data[field]:
            return False, f"{field} is required"
    
    # Validate title
    if len(data['title'].strip()) < 3:
        return False, "Title must be at least 3 characters long"
    
    # Validate price
    try:
        price = float(data['price'])
        if not validate_price(price):
            return False, "Price must be greater than 0"
    except (ValueError, TypeError):
        return False, "Invalid price format"
    
    # Validate SKU
    if not validate_sku(data['sku']):
        return False, "SKU must be at least 3 characters and contain only letters, numbers, hyphens, and underscores"
    
    # Validate category_id
    try:
        category_id = int(data['category_id'])
        if category_id <= 0:
            return False, "Invalid category ID"
    except (ValueError, TypeError):
        return False, "Invalid category ID format"
    
    return True, ""
